pinot-server/src/main/java/org/apache/pinot/server/starter/helix/OffsetBasedConsumptionStatusChecker.java

```diff
@@ -0,0 +1,154 @@
+ * To achieve this, every time status check is called, {@link #haveAllConsumingSegmentsReachedStreamLatestOffset},
+ * list of consuming segments is gathered and then for each segment, we check if segment's latest ingested offset has
+ * reached the latest stream offset. To prevent chasing a moving target, once the latest stream offset is fetched, it
+ * will not be fetched again and subsequent status check calls compare latest ingested offset with the already fetched
+ * stream offset.
+ */
+public class OffsetBasedConsumptionStatusChecker {
+  private static final Logger LOGGER = LoggerFactory.getLogger(OffsetBasedConsumptionStatusChecker.class);
+
+  private final InstanceDataManager _instanceDataManager;
+  private Supplier<Set<String>> _consumingSegmentFinder;
+
+  private Set<String> _alreadyProcessedSegments = new HashSet<>();
+  private Map<String, StreamPartitionMsgOffset> _segmentNameToLatestStreamOffset = new HashMap<>();
+
+  public OffsetBasedConsumptionStatusChecker(InstanceDataManager instanceDataManager, HelixAdmin helixAdmin,
+      String helixClusterName, String instanceId) {
+    this(instanceDataManager, () -> findConsumingSegments(helixAdmin, helixClusterName, instanceId));
+  }
+
+  @VisibleForTesting
+  OffsetBasedConsumptionStatusChecker(InstanceDataManager instanceDataManager,
+      Supplier<Set<String>> consumingSegmentFinder) {
+    _instanceDataManager = instanceDataManager;
+    _consumingSegmentFinder = consumingSegmentFinder;
+  }
+
+  public boolean haveAllConsumingSegmentsReachedStreamLatestOffset() {
+    boolean allSegsReachedLatest = true;
+    Set<String> consumingSegmentNames = _consumingSegmentFinder.get();
+    for (String segName : consumingSegmentNames) {
+      if (_alreadyProcessedSegments.contains(segName)) {
+        continue;
+      }
+      TableDataManager tableDataManager = getTableDataManager(segName);
+      if (tableDataManager == null) {
+        LOGGER.info("TableDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      SegmentDataManager segmentDataManager = tableDataManager.acquireSegment(segName);
+      if (segmentDataManager == null) {
+        LOGGER.info("SegmentDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      if (!(segmentDataManager instanceof LLRealtimeSegmentDataManager)) {
+        // There's a small chance that after getting the list of consuming segment names at the beginning of this method
+        // up to this point, a consuming segment gets converted to a committed segment. In that case status check is
+        // returned as false and in the next round the new consuming segment will be used for fetching offsets.
+        LOGGER.info("Segment {} is already committed. Will check consumption status later", segName);
+        tableDataManager.releaseSegment(segmentDataManager);
```

Comment:
you should release with a try/catch/finally.

Also, can it happen that after the release the segment can move over to be on `OfflineSegmentDataManager`, then we will see an exception for wrong casting. So, grab the offset while you have the lock.

Reply:
Added a try/catch.
For the 2nd part of your comment: at this line of code, we know that segment has already moved over to be an offline segment and then we return false. We don't continue with casting and the rest of the method.